### PR TITLE
Batch `apply_lse_kernel` for `online=True`

### LINKED ISSUE #20: Vectorization memory issues with `PointCloud.apply_lse_kernel` with `online=True`
We've noticed that when `online=True`, the memory performance got worse when `online=False`,
After checking the traceback (attached below), we've found that the 2 netsted `vmaps` in `apply_lse_kernel` vectorize the code such that `n x m` matrix is fully materialized (`n/m` being the number of points the in `PointCloud`).

I have a small fix in [here](https://github.com/michalk8/ott/blob/feature/apply-lse-batch/ott/geometry/pointcloud.py#L129-L178) which uses `online: Optional[int] = None` as a batch size. Then in `apply_lse_kernel`, I use `lax.scan` and within the loop body, the fully vectorized computation for that batch is used - this reduces the memory complexity from `O(n * m) -> O(max(n, m) * batch_size)`.
So far, can only report that I can run `sinkhorn` on a `PointCloud` of shape `(65536, 32768)` using `8516MiB` memory (`ohline=16384`), which previously raised OOM on `16GiB` GPU (Tesla T4).
It took `1157s` to run `sinkhorn` with `epsilon=1e-2` ; will try to do more comprehensive benchmarks 
later.
As for tests, they all pass, except 4:
```bash
FAILED tests/core/sinkhorn_bures_test.py::SinkhornTest::test_bures_point_cloud_ker-batch
FAILED tests/geometry/geometry_pointcloud_apply_test.py::ApplyTest::test_apply_cost_and_kernel
FAILED tests/core/sinkhorn_test.py::SinkhornTest::test_apply_transport_geometry_from_potentials
FAILED tests/core/sinkhorn_test.py::SinkhornTest::test_apply_transport_geometry_from_scalings
```

Think there might be a more efficient approach.

For benchmarking, I've used the same code as in #9 
[jax_online_err.txt](https://github.com/ott-jax/ott/files/8147666/jax_online_err.txt)

### PR DESCRIPTION
**Pull request by michalk8**
As discussed in #20 , this PR fixes ``online`` by batching ``apply_lse_kernel`` and running the fully-vectorized computation on a batch of shape `n * batch_size` or `m * batch_size` (depends of the axis) instead of `n * m`. Minor inefficiency comes from that this approach computed the kernel application for extra `{n,m} % batch_size` points, given that result of each iteration of `jax.lax.scan` must have the same shape.

@LaetitiaPapaxanthos there are 2 points I am unsure how you want to handle:
- [x] ~~shall the backward compatibility be retained and allow `backward=True`? In that case, we'd need to use some default value, that can either be fixed or depend on the number of points (from our benchmarks, value of `1024` seems to work well)~~ **UPDATE**: `online=True` is the same as `online=1024`
- [x] ~~currently, the batch size is the same when using `axis=0` or `axis=1`, but this could be done in axis-specific manner~~ **UPDATE**: kept the same batch size for both axes

TODOs:
- [x] add new tests (and depending on the 1st point above, might need to adjust old tests where `online=True`)
- [x] update docs

closes #20

### CODE REVIEW

**Comment by michalk8** [comment]
There was a bug which caused the coupling/marginals not to match `online=False` (neither tests here or running the tutorial notebook caught it for the particular values of `online`), should be fixed now (added a test for this + 2 more for jitting); locally tests pass (but would be great to enable CI on PRs).
Performance-wise, should even be slightly faster, since `apply_lse_kernel` doesn't do any extra work (`985.29s` for a coupling of shape `(131072, 65536)`). As far as corner-cases, there should be tests for them (batch size of `1`, `n`, `m` and some prime number).

**Comment by marcocuturi** [comment]
Thanks a lot Michal! this is fantastic.
